packages/universal-router-sdk/src/entities/protocols/pancakeswap.ts

```diff
@@ -0,0 +1,403 @@
+    // as a v3 swap is essentially a for of mixedRoute
+    const path = SmartRouter.encodeMixedRouteToPath(
+      { ...route, input: inputAmount.currency, output: outputAmount.currency },
+      trade.tradeType === TradeType.EXACT_OUTPUT,
+    )
+    const amountIn: bigint = SmartRouter.maximumAmountIn(trade, options.slippageTolerance, inputAmount).quotient
+    const amountOut: bigint = SmartRouter.minimumAmountOut(trade, options.slippageTolerance, outputAmount).quotient
+
+    const recipient = routerMustCustody
+      ? ROUTER_AS_RECIPIENT
+      : validateAndParseAddress(options.recipient ?? SENDER_AS_RECIPIENT)
+
+    // similar to encodeV3Swap only we dont need to add a case for signle hop. by using ecodeMixedRoutePath
+    // we can get the parthStr for all cases
+    if (trade.tradeType === TradeType.EXACT_INPUT) {
+      const exactInputSingleParams: ABIParametersType<CommandType.V3_SWAP_EXACT_IN> = [
+        recipient,
+        amountIn,
+        performAggregatedSlippageCheck ? 0n : amountOut,
+        path,
+        payerIsUser,
+      ]
+      planner.addCommand(CommandType.V3_SWAP_EXACT_IN, exactInputSingleParams)
+    } else {
+      const exactOutputSingleParams: ABIParametersType<CommandType.V3_SWAP_EXACT_OUT> = [
+        recipient,
+        amountOut,
+        amountIn,
+        path,
+        payerIsUser,
+      ]
+      planner.addCommand(CommandType.V3_SWAP_EXACT_OUT, exactOutputSingleParams)
+    }
+  }
+}
+
+// encode a mixed route swap, i.e. including both v2 and v3 pools
+async function addMixedSwap(
+  planner: RoutePlanner,
+  trade: SmartRouterTrade<TradeType>,
+  options: PancakeSwapOptions,
+  payerIsUser: boolean,
+  routerMustCustody: boolean,
+  performAggregatedSlippageCheck: boolean,
+): Promise<void> {
+  const isExactIn = trade.tradeType === TradeType.EXACT_INPUT
+
+  for (const route of trade.routes) {
+    const { inputAmount, outputAmount, pools } = route
+    const amountIn: bigint = SmartRouter.maximumAmountIn(trade, options.slippageTolerance, inputAmount).quotient
+    const amountOut: bigint = SmartRouter.minimumAmountOut(trade, options.slippageTolerance, outputAmount).quotient
+
+    // flag for whether the trade is single hop or not
+    const singleHop = pools.length === 1
+
+    const recipient = routerMustCustody
+      ? ROUTER_AS_RECIPIENT
+      : validateAndParseAddress(options.recipient ?? SENDER_AS_RECIPIENT)
+
+    const mixedRouteIsAllV3 = (r: Omit<BaseRoute, 'input' | 'output'>) => {
+      return r.pools.every(SmartRouter.isV3Pool)
+    }
+    const mixedRouteIsAllV2 = (r: Omit<BaseRoute, 'input' | 'output'>) => {
+      return r.pools.every(SmartRouter.isV2Pool)
+    }
+
+    const mixedRouteIsAllStable = (r: Omit<BaseRoute, 'input' | 'output'>) => {
+      return r.pools.every(SmartRouter.isStablePool)
+    }
+
+    // similar to encodeMixedRouteSwap but more simplified where we just continue
+    // as if its a regular v2 or v3 trade
+    if (singleHop) {
```

Comment:
It's just an extra layer of check. Should be the same for both single hop and multi hops